### ISSUE TITLE
.pre-commit directory is not cleaned up

### ISSUE THREAD

**Issue by jtwang** [pre-commit/pre-commit#283]
I noticed at one point that my ~/.pre-commit directory was 3G in size, which seems pretty large.

I asked around and @struys said:

> jtwang: we don't do garbage collection
> so when you get a new version of hooks, the old one stays there
> jtwang: you can also use `pre-commit clean` to remove the files
> rm'ing the directory directly isn't a great idea
> jtwang: mind filing a ticket on pre-commit - https://github.com/pre-commit/pre-commit
> we should add some form of GC based on the last time you used a hook repo

**Comment by asottile** [comment]
This could use access time at best.  I'd say temp watch but that's probably not going to with well on windows.  Also it wouldn't work so well with our bookkeeping database.  I'd also rather not add to the bookkeeping since that makes the upgrade path complicated.

**Comment by asottile** [comment]
atime seems acceptable (at least on Linux) if the disks are mounted with the option enabled.  For more reading, check out https://en.m.wikipedia.org/wiki/Stat_(system_call)#Criticism_of_atime

**Comment by shaleh** [comment]
Why can't the db store a "last used" stamp whenever a pre-commit hook runs? Then you can ask which repo was used least recently and fry it. Worst case is it needs to be recreated.

**Comment by asottile** [comment]
@shaleh It could, I've considered that above:
>  I'd also rather not add to the bookkeeping since that makes the upgrade path complicated.

I'd imagine that could be just a single extra column on the database, upgrade path would be to test the schema and update it if it is out of date.  I guess that wouldn't be the worst.

**Comment by shaleh** [comment]
Unless autoupdate is going to purge old repos I do not see how it would change anything. The out of date check is passed on the sha/tag/whatever.

**Comment by asottile** [comment]
I might take a stab at it this weekend, it's probably easier than I thought and rollback safe

**Comment by shaleh** [comment]
A slightly related question. Say I am a pre-commit user and one of the hooks is misbehaving. How do I go about finding where the hook code is so I can debug it? Looking in ~/.pre-commit I see numerous repos after a some tests and autoupdate calls.

**Comment by asottile** [comment]
@shaleh There isn't a great user-friendly interface for it but:

```
$ sqlite3 ~/.pre-commit/db.db 
SQLite version 3.8.2 2013-12-06 14:53:30
Enter ".help" for instructions
Enter SQL statements terminated with a ";"
sqlite> select * from repos;
https://github.com/pre-commit/pre-commit-hooks.git|cf550fcab3f12015f8676b8278b30e1a5bc10e70|/home/asottile/.pre-commit/repo586Ies
https://github.com/pre-commit/pre-commit.git|8dba3281d5051060755459dcf88e28fc26c27526|/home/asottile/.pre-commit/repo9TbTfw
https://github.com/asottile/reorder_python_imports.git|3d86483455ab5bd06cc1069fdd5ac57be5463f10|/home/asottile/.pre-commit/repot3CqS8
git@github.com:pre-commit/mirrors-scss-lint|3eb13b9647543ad4d6a62c8be8a9131e3b99b96a|/home/asottile/.pre-commit/repoYBDyFK
git@github.com:pre-commit/pre-commit-hooks|516cc9fa72ad09699f2c03ffbd0aa7f60d75b59a|/home/asottile/.pre-commit/repoIPr2Vp
https://github.com/pre-commit/pre-commit-hooks|9ba5af45ce2d29b64c9a348a6fcff5553eea1f2c|/home/asottile/.pre-commit/repoAm4y7d
https://github.com/pre-commit/pre-commit|8dba3281d5051060755459dcf88e28fc26c27526|/home/asottile/.pre-commit/repoVOlKJW
https://github.com/asottile/reorder_python_imports|3d86483455ab5bd06cc1069fdd5ac57be5463f10|/home/asottile/.pre-commit/repoVmqogs
git@github.com:pre-commit/pre-commit-hooks|cf550fcab3f12015f8676b8278b30e1a5bc10e70|/home/asottile/.pre-commit/repoDpb6im
```

**Comment by shaleh** [comment]
I was considering adding a "pre-commit info" or the like but wanted to ask in case I was missing something really simple.

Thanks.

**Comment by sanmai-NL** [comment]
Seems to be an issue still. What are the consequences of removing `~/.pre-commit/`.

**Comment by asottile** [comment]
You can remove `~/.pre-commit` via `pre-commit clean` -- the next invocation of pre-commit will install the required hooks (so your next commit may be slower) but otherwise it is harmless to remove.

**Comment by asottile** [comment]
This has been released as part of [v1.14.0](https://github.com/pre-commit/pre-commit/releases/tag/v1.14.0) 🎉  -- thanks again!
